packages/ifpack2/src/Ifpack2_LocalSparseTriangularSolver_def.hpp

```diff
@@ -358,8 +359,75 @@ initialize ()
     (! G->isFillComplete (), std::runtime_error, "If you call this method, "
      "the matrix's graph must be fill complete.  It is not.");
 
+  if (A_crs_->isUpperTriangular() && htsImpl_.is_null()) {
+    // Reverse the storage for an upper triangular matrix
+    using local_matrix_type = typename crs_matrix_type::local_matrix_type;
+
+    auto Alocal = A_crs_->getLocalMatrix();
+    auto ptr    = Alocal.graph.row_map;
+    auto ind    = Alocal.graph.entries;
+    auto val    = Alocal.values;
+
+    auto numRows = Alocal.numRows(), numCols = Alocal.numCols(), numNnz = Alocal.nnz();
+
+    typename decltype(ptr)::non_const_type  newptr ("ptr", ptr.dimension_0 ());
+    typename decltype(ind)::non_const_type  newind ("ind", ind.dimension_0 ());
+    decltype(val)                           newval ("val", val.dimension_0 ());
+
+    newptr(0) = 0;
+    for (local_ordinal_type row = 0, rowStart = 0; row < numRows; ++row) {
+      auto A_r = Alocal.row(numRows-1 - row);
+
+      auto numEnt = A_r.length;
+      for (local_ordinal_type k = 0; k < numEnt; ++k) {
+        newind(rowStart + k) = numCols-1 - A_r.colidx(numEnt-1 - k);
+        newval(rowStart + k) = A_r.value (numEnt-1 - k);
+      }
+      rowStart += numEnt;
+      newptr(row+1) = rowStart;
+    }
+
+    // Reverse maps
+    using map_type = typename crs_matrix_type::map_type;
+    Teuchos::RCP<map_type> newRowMap, newColMap;
+    {
+      // Reverse row map
+      auto rowMap = A_->getRowMap();
+      auto numElems = rowMap->getNodeNumElements();
+      auto rowElems = rowMap->getNodeElementList();
+
+      Teuchos::Array<global_ordinal_type> newRowElems(rowElems.size());
+      for (size_t i = 0; i < numElems; i++)
+        newRowElems[i] = rowElems[numElems-1 - i];
```

Comment:
I find this way is a lot easier, as the permutation of input and output vectors is done automatically. I think that `apply` may also be slightly faster as instead of `gather vector through import + permute it locally` we only do `gather vector through import into an already permuted storage`. Same with export. Plus, I think the code is cleaner this way as the whole permutation is confined within the `initialize` call.